server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveIndexAction.java

```diff
@@ -0,0 +1,636 @@
+            List<ResolvedIndex> indices = Collections.synchronizedList(new ArrayList<>());
+            List<ResolvedAlias> aliases = Collections.synchronizedList(new ArrayList<>());
+            List<ResolvedDataStream> dataStreams = Collections.synchronizedList(new ArrayList<>());
+            if (localIndices != null) {
+                List<String> resolvedIndexAbstractions = resolveIndexAbstractions(localIndices.indices(), request.indicesOptions, metadata);
+                SortedMap<String, IndexAbstraction> lookup = metadata.getIndicesLookup();
+                for (String s : resolvedIndexAbstractions) {
+                    addIndexAbstraction(s, lookup, indices, aliases, dataStreams);
+                }
+            }
+
+            if (remoteClusterIndices.size() > 0) {
+                final int remoteRequests = remoteClusterIndices.size();
+                final CountDown completionCounter = new CountDown(remoteRequests);
+                final Runnable terminalHandler = () -> {
+                    if (completionCounter.countDown()) {
+                        listener.onResponse(new Response(indices, aliases, dataStreams));
+                    }
+                };
+
+                // make the cross-cluster calls
+                for (Map.Entry<String, OriginalIndices> remoteIndices : remoteClusterIndices.entrySet()) {
+                    String clusterAlias = remoteIndices.getKey();
+                    OriginalIndices originalIndices = remoteIndices.getValue();
+                    Client remoteClusterClient = remoteClusterService.getRemoteClusterClient(threadPool, clusterAlias);
+                    Request remoteRequest = new Request(originalIndices.indices(), originalIndices.indicesOptions());
+                    remoteClusterClient.admin().indices().resolveIndex(remoteRequest, ActionListener.wrap(response -> {
+                        for (ResolvedIndex index : response.indices) {
+                            indices.add(index.copy(RemoteClusterAware.buildRemoteIndexName(clusterAlias, index.getName())));
+                        }
+                        for (ResolvedAlias alias : response.aliases) {
+                            aliases.add(alias.copy(RemoteClusterAware.buildRemoteIndexName(clusterAlias, alias.getName())));
+                        }
+                        for (ResolvedDataStream dataStream : response.dataStreams) {
+                            dataStreams.add(dataStream.copy(RemoteClusterAware.buildRemoteIndexName(clusterAlias, dataStream.getName())));
+                        }
+                        terminalHandler.run();
+                    }, failure -> terminalHandler.run()));
+                }
+            } else {
+                listener.onResponse(new Response(indices, aliases, dataStreams));
+            }
+        }
+
+        private List<String> resolveIndexAbstractions(String[] indices, IndicesOptions indicesOptions, Metadata metadata) {
```

Comment:
This code is largely adapted from `IndicesAndAliasesResolver::resolveIndicesAndAliases`. I saw no clear way to share that code.

Reply:
Do you think that we can add the shared logic to `IndexNameExpressionResolver` and share it between here and `IndicesAndAliasesResolver`?

Reply:
I think that isn't straightforward as the `replaceWildcardsWithAuthorizedIndices()` in `IndicesAndAliasesResolver` does additional filter out non autorized indices and that doesn't happen here.